### ISSUE TITLE
Thrift protocol: incorrect `thriftMethodInDst` behavior when using TThrift

### ISSUE THREAD

**Issue by mirosval** [linkerd/linkerd#1662]
Issue Type:

- [x] Bug report
- [ ] Feature request

**What happened**:

I get an exception `org.apache.thrift.TApplicationException: Internal error processing deleteImage: 'com.twitter.finagle.NoBrokersAvailableException: No hosts are available for /svc/deleteImage, Dtab.base=[/consul=>/#/io.l5d.consul/small-virtualbox;/svc=>/consul], Dtab.local=[]. Remote Info: Not Available'` when I attempt to call my Thrift service.

As you can see, the dentry being resolved is `/svc/deleteImage`, `deleteImage` is a method.

**What you expected to happen**:

According to the documentation [here](https://linkerd.io/config/1.2.1/linkerd/index.html#thrift-protocol) the format should be `/ dstPrefix [/ dest] [/ thriftMethod ] ` but there is nothing in the `dest` segment. I am using Finagle 7.1.0 to construct my client like so:

```scala
new ImageService.FinagledClient(
  Thrift.client
    .withAttemptTTwitterUpgrade
    .newClient(thriftAddress, "image_service_client")
    .toService
)
```

Linkerd 1.2.1 config
```yaml
admin:
  ip: 0.0.0.0
  port: 9990

namers:
- kind: io.l5d.consul
  includeTag: false
  useHealthCheck: false
  setHost: true

routers:
- protocol: http
  label: http
  service:
    responseClassifier:
      kind: io.l5d.http.retryableIdempotent5XX
  identifier:
   kind: io.l5d.methodAndHost
  dtab: |
    /consul => /#/io.l5d.consul/small-virtualbox;
    /local-consul => /$/io.buoyant.http.subdomainOfPfx/service.consul/consul;
    /named-consul => /$/io.buoyant.http.subdomainOfPfx/service.small-virtualbox.consul/consul;
    /svc/*/* => /local-consul | /named-consul;
  servers:
  - port: 80
    ip: 0.0.0.0
  client:
   requeueBudget:
     percentCanRetry: 5.0
- protocol: thrift
  experimental: true
  label: thrift
  dtab: |
    /consul => /#/io.l5d.consul/small-virtualbox;
    /svc => /consul;
  thriftProtocol: binary
  thriftMethodInDst: true
  servers:
  - port: 81
    ip: 0.0.0.0
    thriftFramed: true
  client:
    thriftFramed: true
    attemptTTwitterUpgrade: true
```

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

* HTTP configuration works without problems
* I have tried also ThriftMux on the client and in Linkerd, to no avail
* There is nothing in the logs for thrift, I only get the exception on the client side

**Comment by esbie** [comment]
Thanks @mirosval, we'll take a look

**Comment by amitsaha** [comment]
hey @mirosval (linkerd user here). I am curious why you are doing this:

```
- protocol: thrift
  experimental: true
  label: thrift
  dtab: |
    /consul => /#/io.l5d.consul/small-virtualbox;
    /svc => /consul;
```

instead of:

```
- protocol: thrift
  experimental: true
  label: thrift
  dtab: |
        /svc => /#/io.l5d.consul/.local/thrift;
```

(The `thrift` above is your label in the router configuration)

We are currently running Python thrift services without the `attemptTTwitterUpgrade:True`, but everything works.

**Comment by mirosval** [comment]
@amitsaha There is no particular reason, other than I have tried all possible combinations including the `io.buoyant.http.domainToPathPfx` rewrite namer, which seems to only work for HTTP traffic. So for those experiments it made sense to have 2 rules.

**Comment by amitsaha** [comment]
May be worth trying the approach that has been working for us? (unless you have already tried and failed).

**Comment by mirosval** [comment]
When you do it that way, you need one configuration entry per service. 

I would like to multiplex my Thrift services. I want to have one port for Thrift, same as I have 80 for HTTP and then I want all my Thrift clients to talk to that one port and linkerd would discover & forward my requests to the appropriate one based on the `dest` field of the Thrift message. This should be the Thrift equivalent of HTTP Host header.

**Comment by olix0r** [comment]
To the best of my knowledge, thrift does not encode the service name in the message structure, and so the router has no means to route by service.  This is a deficiency of the thrift protocol.

More modern RPC protocols like Mux or gRPC (which can use thrift encoding for payloads), have a metadata layer that encodes this information, and so routers are easily able to multiplex services based on the headers.

**Comment by mirosval** [comment]
@olix0r My understanding after reading the documentation was that it is indeed the case that Thrift does not include the service name in the message, but TThrift (the Twitter extensions to Thrift) does.

**Comment by mirosval** [comment]
So after further digging, this seems to be fixable with ThriftMux, but you have to explicitly enable muxing protocol like so:

```scala
val client = new ImageService.FinagledClient(
    service = ThriftMux.client.newClient(thriftAddress, "image-service-client").toService,
    protocolFactory = Protocols.multiplex("image-thrift", Protocols.binaryFactory()),
    serviceName = "image-thrift",
    stats = NullStatsReceiver
)
```

**Comment by mirosval** [comment]
I'm sorry I have to reopen this, it seems I was overly excited to see it work, but now I have some additional insights:

So now my error message looks like this:
```
org.apache.thrift.TApplicationException: Internal error processing image-thrift:deleteImage: 'com.twitter.finagle.NoBrokersAvailableException: 
No hosts are available for /svc/image-thrift:deleteImage, 
Dtab.base=[
/thrift/*=>/#/io.l5d.consul/small-virtualbox;
/svc=>/$/io.buoyant.porthostPfx/thrift
], Dtab.local=[]. Remote Info: Not Available'
```
* Good news is that the `image-thrift` is the name of my service, so this is now definitely coming to Linkerd
* Not-so-good news is that it appears that Linkerd is not handling it correctly:
  1. When I set `thriftMethodInDst: true` then both the service name and the method name are appearing like this: `/svc/<service-name>:<method-name>`, I think Linkerd should correctly split it like this: `/svc/<service-name>/<method-name>` and the `method-name` is controlled by the `thriftMethodInDst` parameter.
  2. When I set `thriftMethodInDst: false` then neither the service name nor method name appears
